File: rust_src/50.pow-x-n.rs
/*
 * @lc app=leetcode.cn id=50 lang=rust
 *
 * [50] Pow(x, n)
 */

// @lc code=start
impl Solution {
    pub fn my_pow(mut x: f64, mut n: i32) -> f64 {
		if n == 0
		{
			return 1.0;
		}

		if x == 1.0 || x == 0.0
		{
			return x;
		}
		if x == -1.0
		{
			return if n & 1 == 1 { -1.0 } else { 1.0 };
		}

		if n < -100
		{
			return 0.0;
		}

		if n < 0
		{
			x = 1.0 / x;
			n = -n;
		}

		let mut res = 1.0;
		
		while n != 0
		{
			if n & 1 == 1
			{
				res *= x;
			}
			x *= x;
			n >>= 1;
		}

		return res;
    }
}
// @lc code=end

File: python_src/50.pow-x-n.py
#
# @lc app=leetcode.cn id=50 lang=python3
#
# [50] Pow(x, n)
#
from typing import *


class Solution0:
    def myPow(self, x: float, n: int) -> float:
        if x == 0.: return 0.
        if n < 0: return self.myPow(1 / x, -n)
        if n == 0: return 1.

        e = n >> 1
        tmp = self.myPow(x, e)

        return tmp * tmp * (x if n & 1 == 1 else 1.)
        
# @lc code=start
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0.: return 0.
        if n < 0: return self.myPow(1 / x, -n)
        if n == 0: return 1.

        res = 1.
        prod = x
        while n > 0:
            if n & 1 == 1:
                res *= prod
            prod *= prod
            n >>= 1

        return res
        
# @lc code=end

s = Solution()
print(s.myPow(2., 10))
print(s.myPow(2.1, 3))
print(s.myPow(2., -2))
print(s.myPow(2., -200000000))
